validator/client/wait_for_activation.go

```diff
@@ -37,46 +39,68 @@ func (v *validator) WaitForActivation(ctx context.Context, accountsChangedChan c
 	return v.internalWaitForActivation(ctx, accountsChangedChan)
 }
 
-// internalWaitForActivation recursively waits for at least one active validator key
+// internalWaitForActivation performs the following:
+// 1) While the key manager is empty, subscribe to keymanager changes until some validator keys exist.
+// 2) Open a server side stream for activation events against the given keys.
+// 3) In another go routine, the key manager is monitored for updates and emits an update event on
+// the accountsChangedChan. When an event signal is received, restart the internalWaitForActivation routine.
+// 4) If the stream is reset in error, restart the routine.
+// 5) If the stream returns a response indicating one or more validators are active, exit the routine.
 func (v *validator) internalWaitForActivation(ctx context.Context, accountsChangedChan <-chan [][fieldparams.BLSPubkeyLength]byte) error {
 	ctx, span := trace.StartSpan(ctx, "validator.WaitForActivation")
 	defer span.End()
-
-	// Step 1: Fetch validating public keys.
 	validatingKeys, err := v.km.FetchValidatingPublicKeys(ctx)
 	if err != nil {
 		return errors.Wrap(err, msgCouldNotFetchKeys)
 	}
-
-	// Step 2: If no keys, wait for accounts change or context cancellation.
+	// if there are no validating keys, wait for some
 	if len(validatingKeys) == 0 {
 		log.Warn(msgNoKeysFetched)
 		return v.waitForAccountsChange(ctx, accountsChangedChan)
 	}
 
-	// Step 3: update validator statuses in cache.
-	if err := v.updateValidatorStatusCache(ctx, validatingKeys); err != nil {
+	stream, err := v.validatorClient.WaitForActivation(ctx, &ethpb.ValidatorActivationRequest{
+		PublicKeys: bytesutil.FromBytes48Array(validatingKeys),
+	})
+	if err != nil {
 		return v.retryWaitForActivation(ctx, span, err, "Connection broken while waiting for activation. Reconnecting...", accountsChangedChan)
 	}
 
-	// Step 4: Check and log validator statuses.
-	someAreActive := v.checkAndLogValidatorStatus()
-	if !someAreActive {
-		// Step 6: If no active validators, wait for accounts change, context cancellation, or next epoch.
+	someAreActive := false
+	for !someAreActive {
 		select {
 		case <-ctx.Done():
-			log.Debug("Context closed, exiting WaitForActivation")
+			log.Debug("Context closed, exiting fetching validating keys")
 			return ctx.Err()
 		case <-accountsChangedChan:
 			// Accounts (keys) changed, restart the process.
 			return v.internalWaitForActivation(ctx, accountsChangedChan)
 		default:
-			if err := v.waitForNextEpoch(ctx, v.genesisTime, accountsChangedChan); err != nil {
-				return v.retryWaitForActivation(ctx, span, err, "Failed to wait for next epoch. Reconnecting...", accountsChangedChan)
+			res, err := (stream).Recv() // retrieve from stream one loop at a time
```

Comment:
we should keep investigating ways to get rid of this, and switch with a polling structure